extensions-core/kafka-indexing-service/src/main/java/io/druid/indexing/kafka/KafkaIndexTask.java

```diff
@@ -378,37 +430,50 @@ public boolean apply(Throwable input)
               }
             }
 
-            final long nextOffset = record.offset() + 1;
-            final long endOffset = ioConfig.getEndPartitions().getPartitionOffsetMap().get(record.partition());
-
-            nextOffsets.put(record.partition(), nextOffset);
+            nextOffsets.put(record.partition(), record.offset() + 1);
+          }
 
-            if (nextOffset == endOffset && assignment.remove(record.partition())) {
-              log.info("Finished reading topic[%s], partition[%,d].", record.topic(), record.partition());
-              assignPartitions(consumer, topic, assignment);
-              stillReading = !assignment.isEmpty();
-            }
+          if (nextOffsets.get(record.partition()).equals(endOffsets.get(record.partition()))
+              && assignment.remove(record.partition())) {
+            log.info("Finished reading topic[%s], partition[%,d].", record.topic(), record.partition());
+            assignPartitions(consumer, topic, assignment);
+            stillReading = ioConfig.isPauseAfterRead() || !assignment.isEmpty();
           }
         }
       }
 
-      // Persist pending data.
-      final Committer finalCommitter = committerSupplier.get();
-      driver.persist(finalCommitter);
+      // if we didn't actually generate any segments, don't try to publish as this will throw errors on the overlord
+      if (driver.getActiveSegmentCount() == 0) {
+        log.info("No segments were created by this task");
+        stopGracefully();
```

Comment:
I believe interrupting itself would have been okay-ish here since the only side effect would have been writing an extra log message but I agree - at any rate, the logic here has been reworked so this shouldn't be an issue anymore